azkaban-common/src/main/java/azkaban/scheduler/MissedSchedulesManager.java

```diff
@@ -0,0 +1,213 @@
+    if (project == null) {
+      return false;
+    }
+    List<String> emailRecipients = project.getFlow(flowName).getFailureEmails();
+    if (action.getExecutionOptions().isFailureEmailsOverridden()) {
+      emailRecipients = action.getExecutionOptions().getFailureEmails();
+    }
+
+    try {
+      Future taskFuture = this.missedSchedulesExecutor.submit(
+          new MissedSchedulesOperationTask(
+              missedScheduleTimesInMs,
+              this.emailer,
+              emailRecipients,
+              backExecutionEnabled,
+              action));
+      if (backExecutionEnabled) {
+        LOG.info("Missed schedule task submitted with emailer {} and action {}", emailRecipients, action);
+      }
+      return true;
+    } catch (RejectedExecutionException e) {
+      LOG.error("Failed to add more missed schedules tasks to the thread pool", e);
+      return false;
+    }
+  }
+
+  public void start() {
+    // TODO: add metrics to monitor active threads count, total number of missed schedule tasks, etc.
+    // Create the thread pool
+    this.missedSchedulesExecutor = this.missedSchedulesManagerEnabled
+        ? Executors.newFixedThreadPool(threadPoolSize, new DaemonThreadFactory("azk-missed-schedules-task-pool"))
+        : null;
+    if (this.missedSchedulesManagerEnabled) {
+      LOG.info("Missed Schedule Manager is ready to take tasks.");
+    } else {
+      LOG.info("Missed Schedule Manager is disabled.");
+    }
+  }
+
+  public boolean stop() throws InterruptedException {
+    if (this.missedSchedulesExecutor != null) {
+      this.missedSchedulesExecutor.shutdown();
+      if (!this.missedSchedulesExecutor.awaitTermination(30, TimeUnit.SECONDS)) {
+        this.missedSchedulesExecutor.shutdownNow();
+      }
+    }
+    return true;
+  }
+
+  /**
+   * A MissedSchedule Task is a unit that store the critical information to execute a missedSchedule operation.
+   * A Task is capable to
+   * 1. notice user
+   * 2. execute back execution if enabled feature flag (TODO)
+   * when a missed schedule happens.
+   * */
+  public static class MissedSchedulesOperationTask implements Runnable {
+    private static final Logger LOG = LoggerFactory.getLogger(MissedSchedulesOperationTask.class);
+    private final Emailer emailer;
+    // emailRecipients to notify users the schedules missed
+    private final List<String> emailRecipients;
+    // the email body
+    private final String emailMessage;
+    // back execute action
+    // if user disabled the config, the action could be null.
+    private ExecuteFlowAction executeFlowAction;
+
+    public MissedSchedulesOperationTask(List<Long> missedScheduleTimes,
+        Emailer emailer,
+        List<String> emailRecipients,
+        boolean backExecutionEnabled,
+        ExecuteFlowAction executeFlowAction) {
+      this.emailer = emailer;
+      this.emailRecipients = emailRecipients;
+      this.executeFlowAction = backExecutionEnabled ? executeFlowAction : null;
```

Comment:
I think we'll always need to provide an `executionFlowAction` object regardless of the value of `backExecutionEnabled`. This object cannot be null because we need it to get possible overwrites of failure emails in line 99 right?

Reply:
Talked offline, my bad, emails are resolved outside of the task object so the attribute executeFlowAction can be null.